modules/porous_flow/src/materials/PorousFlowHysteresisOrder.C

```diff
@@ -0,0 +1,195 @@
+                  "PorousFlow_hysteresis_saturation_tps_nodal")
+            : getMaterialPropertyOld<std::array<Real, PorousFlowConstants::MAX_HYSTERESIS_ORDER>>(
+                  "PorousFlow_hysteresis_saturation_tps_qp"))
+{
+  if (_liquid_ph_num >= _dictator.numPhases())
+    paramError("liquid_phase",
+               "The Dictator proclaims that the number of fluid phases is ",
+               _dictator.numPhases(),
+               " while you have foolishly entered ",
+               _liquid_ph_num,
+               " for your liquid phase number.  Remember that indexing starts at 0.  Be aware that "
+               "the Dictator does not tolerate mistakes.");
+  if (_initial_order > PorousFlowConstants::MAX_HYSTERESIS_ORDER)
+    paramError("initial_order",
+               "The initial order must be less than the max order, which is hard-coded to ",
+               PorousFlowConstants::MAX_HYSTERESIS_ORDER);
+  if (_previous_turning_points.size() != _initial_order)
+    paramError("previous_turning_points",
+               "initial_order is ",
+               _initial_order,
+               " so there must be this many previous_turning_points");
+  for (const auto & tp : _previous_turning_points)
+    if (tp < 0.0 || tp > 1.0)
+      paramError("previous_turning_points",
+                 "Each previous_turning_point must lie in the range [0.0, 1.0]");
+  // check that the previous_turning_points are ordered correctly
+  Real previous_tp = 1.0;
+  Real previous_jump = std::numeric_limits<Real>::max();
+  bool drying = true;
+  for (unsigned i = 0; i < _initial_order; ++i)
+  {
+    const Real this_jump = _previous_turning_points[i] - previous_tp;
+    if (drying)
+    {
+      if (this_jump >= 0)
+        paramError("previous_turning_points",
+                   "The previous_turning_points vector, {a, b, c, ...}, must be obey the following "
+                   "order: a < 1; b > a; a < c < b, etc");
+    }
+    else
+    {
+      if (this_jump <= 0)
+        paramError("previous_turning_points",
+                   "The previous_turning_points vector, {a, b, c, ...}, must be obey the following "
+                   "order: a < 1; b > a; a < c < b, etc");
+    }
+    if (std::abs(this_jump) >= previous_jump)
+      paramError("previous_turning_points",
+                 "The previous_turning_points vector, {a, b, c, ...}, must be obey the following "
+                 "order: a < 1; b > a; a < c < b, etc");
+    drying = !drying;
+    previous_tp = _previous_turning_points[i];
+    previous_jump = std::abs(this_jump);
+  }
+}
+
+void
+PorousFlowHysteresisOrder::initQpStatefulProperties()
+{
+  _hys_order[_qp] = _initial_order;
+  for (unsigned i = 0; i < _initial_order; ++i)
+    _hys_sat_tps[_qp].at(i) = _previous_turning_points[i];
+}
+
+void
+PorousFlowHysteresisOrder::computeQpProperties()
+{
+  // NOTE to readers: the computed properties depend on old and older quantities only.  Hence, it
```

Comment:
Could you put this bit in timeStepSetup(). I think it gets called at the start of each timestep, but I would have to check to make sure.

Reply:
I don't think i could utilize timstepsetup() easily because i still have to loop over elements and quadpoints, or elements and nodes in the nodal case.   What do you think?

Reply:
You are probably correct. At least this shouldn't be that expensive compared to everything else!

Reply:
Yea, i agree about the minor expense.  It bugs me right now, but i'll probably forget about it in a few weeks.